src/Interpreters/DatabaseCatalog.cpp

```diff
@@ -121,9 +121,14 @@ TemporaryTableHolder::~TemporaryTableHolder()
 {
     if (id != UUIDHelpers::Nil)
     {
-        auto table = getTable();
-        table->flushAndShutdown();
-        temporary_tables->dropTable(getContext(), "_tmp_" + toString(id));
+        /// Table maybe dropped before by DROP TEMPORARY TABLE
+        /// Thus we need to verify the existence of the table before calling dropTable
+        auto table = temporary_tables->tryGetTable("_tmp_" + toString(id), getContext());
+        if (table && !table->is_dropped)
```

Comment:
It should not happen, AFAIK the only place where a temporary table can be dropped is here

Reply:
Well, I see that it was changed, but I would prefer to keep it the only place if possible

Reply:
I guess we don't really need the lock here because temporary tables are session-local, so no concurrent queries are possible. Also all background operations related to the table are stopped after `flushAndShutdown`

Reply:
But let's wrap `~TemporaryTableHolder()` into `try { ... } catch (...) { tryLogCurrentException(); }` just in case, because `IStorage::drop` may be not trivial for arbitrary engines and it may throw

Reply:
I added try/catch in TemporaryTableHolder destructor.